lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx

```diff
@@ -92,6 +92,47 @@ String IntegrationStrategy::__repr__() const
 }
 
 
+struct IntegrationStrategyCoefficientsPolicy
+{
+  const UnsignedInteger marginalIndex_;
+  const NumericalSample & outputSample_;
+  const NumericalPoint & weights_;
+  const Matrix & designMatrix_;
+  const Indices & addedRanks_;
+  NumericalPoint & alpha_;
+
+  IntegrationStrategyCoefficientsPolicy(const UnsignedInteger marginalIndex,
+					const NumericalSample & outputSample,
+					const NumericalPoint & weights,
+					const Matrix & designMatrix,
+					const Indices & addedRanks,
+					NumericalPoint & alpha)
+    : marginalIndex_(marginalIndex)
+    , outputSample_(outputSample)
+    , weights_(weights)
+    , designMatrix_(designMatrix)
+    , addedRanks_(addedRanks)
+    , alpha_(alpha)
+  {
+    // Nothing to do
+  }
+
+  inline void operator()( const TBB::BlockedRange<UnsignedInteger> & r ) const
+  {
+    for (UnsignedInteger j = r.begin(); j != r.end(); ++j)
+      {
+	for (UnsignedInteger i = 0; i < outputSample_.getSize(); ++i)
+	  {
+	    const UnsignedInteger indexAdded(addedRanks_[j]);
+	    // The integral is approximated by a weighted sum of f * Phi
+	    alpha_[j] += weights_[i] * outputSample_[i][marginalIndex_] * designMatrix_(i, indexAdded);
```

Comment:
@dbarbier I first wrote a wrong code where the i and j loops were interchanged, then I fixed the mess but leaving things in the innermost loop by mistake. Concerning the linear algebra aspects, the situation is more involved than it looks at first sight. For some of our applications, the database can be VERY large, say 1e6 points in dimension 50, so it is not a good idea to store things twice. For the algorithm, only the weighted version of the output sample is needed, but the user may want to access to the output sample, so both the weighted and unweighted versions of the output sample are needed. I chose to keep only the unweighted sample and to recompute the weighted version on the fly, but other options are possible, such as a threshold below which things are stored and above which they are recomputed on the fly. Do you have a better idea to cope with these constraints?